src/test/java/org/broadinstitute/hellbender/utils/baq/BAQUnitTest.java

```diff
@@ -1,26 +1,230 @@
+
+        // raw base qualities are low -- but they shouldn't be capped
+        params.add(new BAQTest(-3, 9999993L, "4=13X2=3X1=4X2=4X1=2X",
+                refString3,
+                "CCACGCTTGGCAAAGTTTTCCGTACGTTTAGCCGAG",
+                "33'/(7+270&4),(&&-)$&,%7$',-/61(,6?8",
+                "33'/(7+270&4),(&&-)$&,%7$',-/61(,6?8",
+                rds3));
+
+        List<Object[]> params2 = new ArrayList<>();
+        for (BAQTest x : params) params2.add(new Object[]{x});
+        return params2.toArray(new Object[][]{});
+    }
+
+
+    @Test(dataProvider = "data")
+    public void testBAQWithProvidedReference(BAQTest test) {
+        if (test.refBases != null) {
+            testBAQ(test, false, null);
+        }
+    }
+
+    @Test(dataProvider = "data")
+    public void testBAQWithCigarAndRefLookup(BAQTest test) {
+        if (test.cigar != null) {
+            testBAQ(test, true, test.rds);
+        }
+    }
+
+    @Test
+    public void testBAQQualRange() {
+        BAQ baq = new BAQ(1.0e-3, 0.1, 7, (byte) 4);         // matches current samtools parameters
+        final byte ref = (byte) 'A';
+        final byte alt = (byte) 'A';
+
+        for (int i = 0; i <= SAMUtils.MAX_PHRED_SCORE; i++) {
+            Assert.assertTrue(baq.calcEpsilon(ref, alt, (byte) i) >= 0.0, "Failed to get baq epsilon range");
+        }
+    }
+
+    @Test
+    public void testBAQOverwritesExistingTagWithNull() {
+        final File referenceFile = new File(hg19_chr1_1M_Reference);
+        final ReferenceDataSource rds = new ReferenceFileSource(referenceFile);
+
+        // create a read with a single base off the end of the contig, which cannot be BAQed
+        final GATKRead read = ArtificialReadUtils.createArtificialRead(createHeader(), "foo", 0, rds.getSequenceDictionary().getSequence("1").getSequenceLength() + 1, 1);
+        read.setBases(new byte[]{(byte) 'A'});
+        read.setBaseQualities(new byte[]{(byte) 20});
+        read.setCigar("1M");
+        read.setAttribute("BQ", "A");
+
+        // try to BAQ and tell it to RECALCULATE AND ADD_TAG
+        final BAQ baq = new BAQ(1.0e-3, 0.1, 7, (byte) 4);
+        baq.baqRead(read, rds, BAQ.CalculationMode.RECALCULATE, BAQ.QualityMode.ADD_TAG);
+
+        // did we remove the existing tag?
+        Assert.assertFalse(read.hasAttribute("BQ"));
+    }
+
+    private void testBAQ(BAQTest test, boolean lookupWithFasta, ReferenceDataSource rds) {
+        final int bandWidth = 7;
+        final BAQ baqHMM = new BAQ(1.0e-3, 0.1, bandWidth, (byte) 4);         // matches current samtools parameters
+
+        final GATKRead read = test.createRead();
+        final BAQ.BAQCalculationResult result;
+        if (lookupWithFasta && test.cigar != null && rds != null) {
+            result = baqHMM.calcBAQFromHMM(read, rds);
+        } else {
+            result = baqHMM.calcBAQFromHMM(read, test.refBases.getBytes(), test.refOffset);
```

Comment:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.baq\.BAQUnitTest\.testBAQ\(BAQUnitTest$BAQTest, boolean, ReferenceDataSource\): String\.getBytes\(\)